subsys/bluetooth/controller/ll_sw/ull_llcp_conn.c

```diff
@@ -0,0 +1,766 @@
+			break;
+		default:
+			/* Unknown role */
+			LL_ASSERT(0);
+		}
+	}
+}
+
+static void lp_cu_send_conn_update_ind(struct ull_cp_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+{
+	if (!tx_alloc_is_available()) {
+		ctx->state = LP_CU_STATE_WAIT_TX_CONN_UPDATE_IND;
+	} else {
+		/* TODO(thoh): Hardcoded instant delta +6 */
+		ctx->data.cu.instant = lp_event_counter(conn) + 6;
+		lp_cu_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_CONN_UPDATE_IND);
+		ctx->rx_opcode = 0xFF; /* TODO(thoh): Hmm */
+		ctx->state = LP_CU_STATE_WAIT_INSTANT;
+	}
+}
+
+static void lp_cu_st_idle(struct ull_cp_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+{
+	/* TODO */
+	switch (evt) {
+	case LP_CU_EVT_RUN:
+		lp_cu_send_conn_param_req(conn, ctx, evt, param);
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void lp_cu_st_wait_tx_conn_param_req(struct ull_cp_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+{
+	switch (evt) {
+	case LP_CU_EVT_RUN:
+		lp_cu_send_conn_param_req(conn, ctx, evt, param);
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void lp_cu_st_wait_rx_conn_param_rsp(struct ull_cp_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+{
+	struct pdu_data *pdu = (struct pdu_data *) param;
+
+	switch (evt) {
+	case LP_CU_EVT_CONN_PARAM_RSP:
+		rr_set_incompat(conn, INCOMPAT_RESERVED);
+		lp_cu_send_conn_update_ind(conn, ctx, evt, param);
+		break;
+	case LP_CU_EVT_UNKNOWN:
+		rr_set_incompat(conn, INCOMPAT_RESERVED);
+		lp_cu_send_conn_update_ind(conn, ctx, evt, param);
+		break;
+	case LP_CU_EVT_REJECT:
+		rr_set_incompat(conn, INCOMPAT_NO_COLLISION);
+		ctx->data.cu.error = pdu->llctrl.reject_ext_ind.error_code;
+		lp_cu_complete(conn, ctx, evt, param);
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void lp_cu_st_wait_tx_conn_update_ind(struct ull_cp_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+{
+	switch (evt) {
+	case LP_CU_EVT_RUN:
+		lp_cu_send_conn_update_ind(conn, ctx, evt, param);
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void lp_cu_st_wait_rx_conn_update_ind(struct ull_cp_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+{
+	struct pdu_data *pdu = (struct pdu_data *) param;
+
+	switch (evt) {
+	case LP_CU_EVT_CONN_UPDATE_IND:
+		pdu_decode_conn_update_ind(ctx, param);
+		ctx->state = LP_CU_STATE_WAIT_INSTANT;
+		break;
+	case LP_CU_EVT_UNKNOWN:
+		ctx->data.cu.error = BT_HCI_ERR_UNSUPP_REMOTE_FEATURE;
+		lp_cu_complete(conn, ctx, evt, param);
+		break;
+	case LP_CU_EVT_REJECT:
+		ctx->data.cu.error = pdu->llctrl.reject_ext_ind.error_code;
+		lp_cu_complete(conn, ctx, evt, param);
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void lp_cu_check_instant(struct ull_cp_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+{
+	uint16_t event_counter = lp_event_counter(conn);
+	if (((event_counter - ctx->data.cu.instant) & 0xFFFF) <= 0x7FFF) {
```

Comment:
This construct is also used at other places, and takes a minute to understand what is being done. Can we replace this with a macro or inline function?